### PR TITLE
Add Manifest to state.Contract

### LINKED ISSUE #923: Add Manifest to ContractState
See https://github.com/neo-project/neo/pull/766 for details.
Also after this we are able to finalize part of #920 with checking transaction's witness (marked with TODO in ` pkg/core/interop/runtime/witness.go`)

### PR DESCRIPTION
**Pull request by fyrchik**
Closes #923 .
In C# implementation manifest is marshaled to binary via JSON. I am not sure if `ContractState` will not be used in internode comunication, so didn't change it.

### CODE REVIEW

**Comment by roman-khimov** [comment]
> In C# implementation manifest is marshaled to binary via JSON.

Weird. Luckily we don't have thousands of contracts.

At the moment contract states are not considered for StateRoot calculations, although they probably should be.

File: pkg/core/interop_neo.go [hunk 1]
```diff
@@ -317,30 +288,46 @@ func contractIsPayable(ic *interop.Context, v *vm.VM) error {
...
+		newcontract.ID = contract.ID
+		if err := ic.DAO.PutContractState(newcontract); err != nil {
+			return err
+		}
+		if err := ic.DAO.DeleteContractState(contract.ScriptHash()); err != nil {
+			return err
+		}
+	}
+	if contract.HasStorage() {
+		// TODO store items by ID
```
Comment (roman-khimov) [inline pkg/core/interop_neo.go:322]: Issue #?

File: pkg/vm/vm.go [hunk 2]
```diff
@@ -136,6 +136,12 @@ func (v *VM) SetGasLimit(max util.Fixed8) {
 	v.gasLimit = max
 }
 
+// AddGas consumes specified amount of gas.
```
Comment (roman-khimov) [inline pkg/vm/vm.go:139]: How can I know what it returns from this description?

File: pkg/core/interop_neo.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/interop_neo.go:322]: It actually is an important one as it directly affects state calculations.

File: pkg/core/interop_neo.go [hunk 3]
```diff
@@ -231,48 +232,23 @@ func createContractStateFromVM(ic *interop.Context, v *vm.VM) (*state.Contract,
...
+	if !v.AddGas(util.Fixed8(interopGasRatio * (len(script) + len(manifestBytes)))) {
+		return nil, errors.New("gas limit exceeded")
+	}
+	var m manifest.Manifest
+	r := io.NewBinReaderFromBuf(manifestBytes)
+	m.DecodeBinary(r)
+	if r.Err != nil {
+		return nil, r.Err
+	}
+	return &state.Contract{
```
Comment (roman-khimov) [inline pkg/core/interop_neo.go:248]: And what's the ID for the new contract?

**Comment by roman-khimov** [comment]
And we also need to change our contract deployment routines now.

File: pkg/vm/vm.go (hunk 2 above)
Comment (fyrchik) [reply pkg/vm/vm.go:139]: Fixed.

File: pkg/core/interop_neo.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/interop_neo.go:322]: https://github.com/nspcc-dev/neo-go/issues/1037

File: pkg/core/interop_neo.go (hunk 3 above)
Comment (fyrchik) [reply pkg/core/interop_neo.go:248]: Forget this part. It is set differently in `contractCreate` and `contractMigrate`

File: cli/smartcontract/smart_contract.go [hunk 4]
```diff
@@ -646,12 +652,16 @@ func contractDeploy(ctx *cli.Context) error {
 		return cli.NewExitError(err, 1)
 	}
 
-	txScript, err := request.CreateDeploymentScript(avm, &conf.Contract)
+	m := conf.ToManifest(avm)
+	w := io.NewBufBinWriter()
+	m.EncodeBinary(w.BinWriter)
+	rawManifest := w.Bytes()
+	txScript, err := request.CreateDeploymentScript(avm, rawManifest)
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:659]: Why `rawManifest`? I think it'd be more convenient to pass normal manifest. Maybe even as Serializable. And we could avoid some excessive buffering that way.

Comment (fyrchik) [reply cli/smartcontract/smart_contract.go:659]: Initally that was to not marshal manifest twice. But `contractDeploy` won't be called frequently, so this is not a big deal I suppose.

File: pkg/core/state/contract.go [hunk 5]
```diff
@@ -1,52 +1,38 @@
...
-	Script      []byte
-	ParamList   []smartcontract.ParamType
-	ReturnType  smartcontract.ParamType
-	Properties  smartcontract.PropertyState
-	Name        string
-	CodeVersion string
-	Author      string
-	Email       string
-	Description string
+	ID       uint32
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/state/contract.go]: Yes, somehow missed this. Fixed.

File: cli/smartcontract/smart_contract.go (hunk 4 above)
Comment (fyrchik) [reply cli/smartcontract/smart_contract.go:659]: Fixed.

File: pkg/core/native/interop.go [hunk 6]
```diff
@@ -26,9 +26,9 @@ func Deploy(ic *interop.Context, _ *vm.VM) error {
 		}
 
 		cs := &state.Contract{
-			Script:     md.Script,
-			ParamList:  params,
-			ReturnType: md.Manifest.ABI.EntryPoint.ReturnType,
+			ID:       md.ServiceID,
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/interop.go]: Fixed.

File: pkg/interop/contract/contract.go [hunk 7]
```diff
@@ -60,12 +38,12 @@ func Create(
 	return Contract{}
 }
 
-// Migrate migrates calling contract (that is the one that calls Migrate) to
+// Update migrates calling contract (that is the one that calls Migrate) to
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/interop/contract/contract.go]: Fixed.

File: pkg/core/interops.go [hunk 8]
```diff
@@ -104,12 +105,6 @@ var systemInterops = []interop.Function{
 }
 
 var neoInterops = []interop.Function{
-	{Name: "Neo.Contract.Create", Func: contractCreate, Price: 0},
-	{Name: "Neo.Contract.Destroy", Func: contractDestroy, Price: 1},
-	{Name: "Neo.Contract.GetScript", Func: contractGetScript, Price: 1},
-	{Name: "Neo.Contract.GetStorageContext", Func: contractGetStorageContext, Price: 1},
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/interops.go:110]: Done.

**Pull request merged**